luaui/Include/blueprint_substitution/logic.lua

```diff
@@ -0,0 +1,489 @@
+    if not myTeamID then return nil end
+
+    local teamUnits = Spring.GetTeamUnits(myTeamID)
+    if not teamUnits or #teamUnits == 0 then return nil end
+
+    -- Check for commander units to determine side
+    for _, unitID in ipairs(teamUnits) do
+        local unitDefID = Spring.GetUnitDefID(unitID)
+        if unitDefID then
+            local unitDef = UnitDefs[unitDefID]
+            if unitDef and unitDef.name then
+                local unitName = unitDef.name:lower()
+
+                -- Use the values (prefixes) from SIDES, not the keys
+                for side, prefix in pairs(BlueprintSubLogic.SIDES) do
+                    if unitName:find(prefix .. "com") then
+                        return prefix
+                    end
+                end
+            end
+        end
+    end
+
+    return nil -- No side detected - this will cause failure instead of defaulting
+end
+
+-- Function to analyze blueprint source sides
+function BlueprintSubLogic.analyzeBlueprintSides(blueprint)
+    if not blueprint or not blueprint.units then
+        return {
+            unitCount = 0,
+            primarySourceSide = nil,
+            numSourceSides = 0,
+            sourceSideInfo = "Empty",
+            sideCounts = {}
+        }
+    end
+
+    local sourceSides = {}
+    local unitCount = #blueprint.units
+    local primarySourceSide = nil
+    local sideCounts = {}
+
+    for _, unit in ipairs(blueprint.units) do
+        local unitName = UnitDefs[unit.unitDefID] and UnitDefs[unit.unitDefID].name:lower()
+        if unitName then
+            local side = BlueprintSubLogic.getSideFromUnitName(unitName)
+            if side then
+                sourceSides[side] = true
+                sideCounts[side] = (sideCounts[side] or 0) + 1
+            end
+        end
+    end
+
+    local numSourceSides = 0
+    local maxCount = 0
+    for side, count in pairs(sideCounts) do
+        numSourceSides = numSourceSides + 1
+        if count > maxCount then
+            primarySourceSide = side
+            maxCount = count
+        end
+    end
+
+    if not primarySourceSide and unitCount > 0 then
+        primarySourceSide = "Mixed/Unknown"
+    end
+
+    local sourceSideInfo = primarySourceSide or "Empty"
+    if numSourceSides > 1 then
+        sourceSideInfo = sourceSideInfo .. " (Mixed)"
+    end
+
+    return {
+        unitCount = unitCount,
+        primarySourceSide = primarySourceSide,
+        numSourceSides = numSourceSides,
+        sourceSideInfo = sourceSideInfo,
+        sideCounts = sideCounts
+    }
+end
+
+function BlueprintSubLogic.generateEquivalentUnits()
+    Spring.Log("BlueprintSubLogic", "info", "Generating equivalent units...")
+    local categoryUnits = BlueprintSubLogic.categoryUnits
+    local equivalentUnits = BlueprintSubLogic.equivalentUnits
+
+    for k in pairs(equivalentUnits) do equivalentUnits[k] = nil end -- Clear before generating
+
+    local mappingCount = 0
```

Comment:
would love functional constructs here, is that a thing in lua? Map/reduce/filter?

Reply:
check these out
https://github.com/beyond-all-reason/Beyond-All-Reason/blob/1d8555f625ab247632c0c1a86be239b1b6c4afb7/common/tablefunctions.lua#L311-L382

Reply:
Nice. Ty

Reply:
That helped a lot. I am now done refactoring if you want to take another look. We were able to pre-munge a "master building list" with our equivalences mapped out for every possible faction permutation. This simplified everything and let us keep it more map-reducey/flat.